### PR TITLE
Fix issue #977

### PR DESCRIPTION
**Pull request by kettenis**
Swap the rows corresponding to cros--polarization products in
DATA, MODEL_DATA, CORRECTED_DATA, WEIGHT, WEIGHT_SPECTRUM, SIGMA,
SIGMA_SPECTRUM, FLAG and FLAG_CATEGORY columns if necessary when
swapping antennas for a baseline.

### CODE REVIEW

**Comment by dpetry** [comment]
Hi Mark, thanks for working on concat. This is quite a massive change. What did you do in terms of testing?

**Comment by kettenis** [comment]
I verified that the following CASA tests still pass:
test_concat
test_importfitsidi
test_testconcat
test_virtualconcat
Unfortunately I have no easy way to run the entire CASA testsuite.

Michael Janssen, who reported the issue to me, gave me some data that triggers the issue. I have verified that this pull request fixes the issue in that data. Unfortunately that data can't be shared outside the EHT collaboration at this point.

**Comment by dpetry** [comment]
For ALMA we haven't noticed this because arrays stay frozen during full pol. sessions and we don't concatenate more than one session typically.
Anyway, your testing looks complete. Let me just have another look at the code ...

**Comment by dpetry** [comment]
P.S.: could you also run the alma-m100-analysis-regression and the leo2pt_regression?

File: ms/MSOper/MSConcat.cc [hunk 1]
```diff
@@ -668,7 +672,27 @@ IPosition MSConcat::isFixedShape(const TableDesc& td) {
...
+  vector<Int> polSwap;
+
   for (uInt r = 0; r < otherRows; r++) {
+    uInt d = otherDDId(r);
+    uInt p = otherDDCols.polarizationId()(otherDDId(r));
+    if (p != polId) {
+      const Matrix<Int> &products = otherPolCols.corrProduct()(p);
+      polSwap.resize(products.shape()(1));
+      for (Int i = 0; i < products.shape()(1); i++) {
+	for (Int j = 0; j < products.shape()(1); j++)
```
Comment (dpetry) [inline ms/MSOper/MSConcat.cc]: The second "for" should be indented.

Comment (dpetry) [reply ms/MSOper/MSConcat.cc]: And it should have a pair of brackets!

Comment (dpetry) [reply ms/MSOper/MSConcat.cc]: In fact, could you please add a comment to the code to explain what exactly is supposed to happen here?
I'm not sure I understand.

File: ms/MSOper/MSConcat.cc [hunk 2]
```diff
@@ -780,13 +834,92 @@ IPosition MSConcat::isFixedShape(const TableDesc& td) {
...
+	}
+	otherSigma.put(r, swappedSigma);
+	if (copySgSp) {
+	  datShape.assign(otherSigmaSp.shape(r).asVector());
+	  Matrix<Float> swappedSigmaSp(datShape[0], datShape[1]);
+	  for (Int p = 0; p < datShape[0]; p++) {
+	    swappedSigmaSp.row(p) = (Matrix<Float>(otherSigmaSp(r))).row(polSwap[p]);
+	  }
+	  otherSigmaSp.put(r, swappedSigmaSp);
+	}
```
Comment (dpetry) [inline ms/MSOper/MSConcat.cc]: This whole block needs to be indented.

File: ms/MSOper/MSConcat.cc [hunk 3]
```diff
@@ -780,13 +834,92 @@ IPosition MSConcat::isFixedShape(const TableDesc& td) {
...
+      }
+    }
+
+    if(doConjugateVis){
+      Vector<Int> datShape=otherFlag.shape(r).asVector();
+      Matrix<Bool> swappedFlag(datShape[0], datShape[1]);
+      for (Int p = 0; p < datShape[0]; p++) {
+	swappedFlag.row(p) = (Matrix<Bool>(otherFlag(r))).row(polSwap[p]);
```
Comment (dpetry) [inline ms/MSOper/MSConcat.cc]: Indent!

File: ms/MSOper/MSConcat.cc [hunk 4]
```diff
@@ -780,13 +834,92 @@ IPosition MSConcat::isFixedShape(const TableDesc& td) {
...
+      }
+      otherFlag.put(r, swappedFlag);
+      if (copyFlagCat) {
+	datShape.assign(otherFlagCat.shape(r).asVector());
+	Cube<Bool> swappedFlagCat(datShape[0], datShape[1], datShape[2]);
+	for (Int p = 0; p < datShape[0]; p++) {
+	  swappedFlagCat.yzPlane(p) = (Cube<Bool>(otherFlagCat(r))).yzPlane(polSwap[p]);
+	}
+	otherFlagCat.put(r, swappedFlagCat);
```
Comment (dpetry) [inline ms/MSOper/MSConcat.cc]: Indent.

File: ms/MSOper/MSConcat.cc [hunk 5]
```diff
@@ -1306,8 +1439,28 @@ IPosition MSConcat::isFixedShape(const TableDesc& td) {
...
+      const Matrix<Int> &products = otherPolCols.corrProduct()(p);
+      polSwap.resize(products.shape()(1));
+      for (Int i = 0; i < products.shape()(1); i++) {
+	for (Int j = 0; j < products.shape()(1); j++)
+	  if (products(0, i) == products(1, j) &&
+	      products(1, i) == products(0, j)) {
+	    polSwap[i] = j;
+	    break;
+	  }
+      }
```
Comment (dpetry) [inline ms/MSOper/MSConcat.cc]: Indenting plus the "for" is missing the brackets!

File: ms/MSOper/MSConcat.cc [hunk 6]
```diff
@@ -1375,6 +1528,7 @@ IPosition MSConcat::isFixedShape(const TableDesc& td) {
       Vector<Int> datShape;
       Matrix<Complex> reversedData;
       Matrix<Float> reversedFloatData;
+      Matrix<Complex> swappedData;
       if(doFloatData){
 	datShape=otherFloatData.shape(r).asVector();
 	reversedFloatData.resize(datShape[0], datShape[1]);
```
Comment (dpetry) [inline ms/MSOper/MSConcat.cc]: Indentation. How come this got lost?

File: ms/MSOper/MSConcat.cc [hunk 7]
```diff
@@ -1465,16 +1647,82 @@ IPosition MSConcat::isFixedShape(const TableDesc& td) {
...
+	}
+	thisSigma.put(curRow, swappedSigma*sScale);
+	if (copySgSp) {
+	  datShape.assign(otherSigmaSp.shape(r).asVector());
+	  Matrix<Float> swappedSigmaSp(datShape[0], datShape[1]);
+	  for (Int p = 0; p < datShape[0]; p++) {
+	    swappedSigmaSp.row(p) = (Matrix<Float>(otherSigmaSp(r))).row(polSwap[p]);
+	  }
+	  thisSigmaSp.put(curRow, swappedSigmaSp*sScale);
+	}
```
Comment (dpetry) [inline ms/MSOper/MSConcat.cc]: Indentation!

File: ms/MSOper/MSConcat.cc [hunk 8]
```diff
@@ -1465,16 +1647,82 @@ IPosition MSConcat::isFixedShape(const TableDesc& td) {
...
+      }
+      else {
+	thisWeight.put(curRow, otherWeight(r)*itsWeightScale);
+	if (copyWtSp)
+	  thisWeightSp.put(curRow, otherWeightSp(r)*itsWeightScale);
+	thisSigma.put(curRow, otherSigma(r)*sScale);
+	if (copySgSp)
+	  thisSigmaSp.put(curRow, otherSigmaSp(r)*sScale);
```
Comment (dpetry) [inline ms/MSOper/MSConcat.cc]: Indentation!

File: ms/MSOper/MSConcat.cc [hunk 9]
```diff
@@ -1465,16 +1647,82 @@ IPosition MSConcat::isFixedShape(const TableDesc& td) {
...
+	}
+	thisSigma.put(curRow, swappedSigma);
+	if (copySgSp) {
+	  datShape.assign(otherSigmaSp.shape(r).asVector());
+	  Matrix<Float> swappedSigmaSp(datShape[0], datShape[1]);
+	  for (Int p = 0; p < datShape[0]; p++) {
+	    swappedSigmaSp.row(p) = (Matrix<Float>(otherSigmaSp(r))).row(polSwap[p]);
+	  }
+	  thisSigmaSp.put(curRow, swappedSigmaSp);
+	}
```
Comment (dpetry) [inline ms/MSOper/MSConcat.cc]: dto.

File: ms/MSOper/MSConcat.cc [hunk 10]
```diff
@@ -1465,16 +1647,82 @@ IPosition MSConcat::isFixedShape(const TableDesc& td) {
...
+      }
+      else{
+	thisWeight.put(curRow, otherWeight, r);
+	if (copyWtSp) thisWeightSp.put(curRow, otherWeightSp, r);
+	thisSigma.put(curRow, otherSigma, r);
+	if (copySgSp) thisSigmaSp.put(curRow, otherSigmaSp, r);
```
Comment (dpetry) [inline ms/MSOper/MSConcat.cc]: Indent ...

File: ms/MSOper/MSConcat.cc [hunk 11]
```diff
@@ -1497,8 +1745,26 @@ IPosition MSConcat::isFixedShape(const TableDesc& td) {
     thisExposure.put(curRow, otherExposure, r);
     thisTimeCen.put(curRow, otherTimeCen, r);
     thisArrayId.put(curRow, otherArrayId, r);
-    thisFlag.put(curRow, otherFlag, r);
-    if (copyFlagCat) thisFlagCat.put(curRow, otherFlagCat, r);
+    if(doConjugateVis){
+      Vector<Int> datShape=otherFlag.shape(r).asVector();
+      Matrix<Bool> swappedFlag(datShape[0], datShape[1]);
+      for (Int p = 0; p < datShape[0]; p++) {
+	swappedFlag.row(p) = (Matrix<Bool>(otherFlag(r))).row(polSwap[p]);
```
Comment (dpetry) [inline ms/MSOper/MSConcat.cc]: Indent!

File: ms/MSOper/MSConcat.cc [hunk 12]
```diff
@@ -1497,8 +1745,26 @@ IPosition MSConcat::isFixedShape(const TableDesc& td) {
...
+      }
+      thisFlag.put(curRow, swappedFlag);
+      if (copyFlagCat) {
+	datShape.assign(otherFlagCat.shape(r).asVector());
+	Cube<Bool> swappedFlagCat(datShape[0], datShape[1], datShape[2]);
+	for (Int p = 0; p < datShape[0]; p++) {
+	  swappedFlagCat.yzPlane(p) = (Cube<Bool>(otherFlagCat(r))).yzPlane(polSwap[p]);
+	}
+	thisFlagCat.put(curRow, swappedFlagCat);
```
Comment (dpetry) [inline ms/MSOper/MSConcat.cc]: Indent.

**Comment by dpetry** [comment]
Hi Mark, do you agree with the suggested changes and additional tests?

**Comment by kettenis** [comment]
Hi Dirk,
I am currently traveling, and haven't had a chance to look at your comments yet. I did try running the alma-m100-analysis-regression test but that one fails even without my fix. I did my testing in the old casa tree though that hasn't been updated since mid-october.  So I'll try again with the casa6 when I get the chance.

**Comment by kettenis** [comment]
Hi Dirk,
I've made the requiested changes to the polswap detection code in my local tree and added the comment:
```
    // Determine whether we need to swap rows in the visibility matrix          
    // if we change the order of the antennas.  This is done by                 
    // creating a mapping that makes sure the receptor numbers remain           
    // correct when the antennas are swapped.                     
```              
Not sure that is very clear, but it tries to convey that the mapping is based on receptor numbers.
As for the indentation problems: I think some of those can be traced back to the mix of tabs and spaces which makes some lines appear to be lined up in the diff output when they aren't in the actual code. There are some cases where an if clause and subsequent conditional statement occur on the same line, but those follow the style of the code that was already there.  I can fix those if you think that is clearer.

I'm still struggling to run the requested tests in the casa6 environment. Do you know how to do this?

**Comment by dpetry** [comment]
Hi, OK, I think, this is is a pure casacore change which does not relate to Python in any way, it would be sufficient if you run the requested tests in the CASA5 environment, if that works better for you.

**Comment by kettenis** [comment]
Hi Dirk,

I can't get the alma-m100-analysis test to run successfully on our developmentsystem, with or without my changes.  It always fails with:

> Exception: Results are different from expectations by more than 1.0 percent.

The reported parameters are identical though, so as far as I can determine the change does not affect the outcome of the test in any way.

The requested changes are now pushed to the branch. Let me know if any further changes are needed.

**Comment by dpetry** [comment]
OK, I approve the changes. Now one check fails. Looks like a test infrastructure problem.

**Comment by dpetry** [comment]
@tammojan , can you judge whether the failing checks for this issue are serious?
To me it looks like a test infrastructure problem.

**Comment by tammojan** [comment]
The failing tests are due to a misinstalled wget on some test systems, not related to this issue.

**Pull request merged**